src/main/java/org/javarosa/xpath/analysis/XPathAnalyzer.java

```diff
@@ -51,17 +53,13 @@ public TreeReference getOriginalContextRef() {
 
     protected void requireOriginalContext(TreeReference forReference) throws AnalysisInvalidException{
         if (getOriginalContextRef() == null) {
-            throw new AnalysisInvalidException("No original context ref was available when " +
```

Comment:
@ctsims I agree, but the reason I made this change was that a while back when I was profiling this I discovered that the creation of the exception objects was actually eating into the performance of the analysis by a noticeable amount. So saving that time seemed more important that preserving the exception messages

Reply:
Aha, good catch. 

Seems like a bit of a shame to lose the capacity all together. It would be possible to either pre-package static, canned error messages (like "no context ref was available why trying to analyse a relative reference") or to do a debug dispatch like

```
public static final bool ANALYSIS_DEBUG = false;

public static AnalysisInvalidException ProduceInvalidException(String message) {
    if(ANALYSIS_DEBUG) { return new AnalysisInvalidException(message);}
    else { return STATIC_INSTANCE; }
}
```

so that in tests or in debugging sessions people can dig into things a bit more, but agree that the performance considerations should be weighed ahead of that value in terms of priority.